docs/source/configuration.rst

```diff
@@ -8,27 +8,98 @@ unwieldy and lead to an overly cumbersome command. It also becomes difficult to
-``/home/my_user/projects/my_project``. ``tartufo`` will look for the
-configuration files first in this directory, then in ``/home/my_user/projects/``,
-then in ``/home/my_user``, etc.
-
-Within these files, ``tartufo`` will look for a section labeled
-``[tool.tartufo]`` to find its configuration, and will load all items from there
-just as though they had been specified on the command line. This file must be
+configuration file!
+
+You can `tell tartufo what config file to use
+<usage.html#cmdoption-tartufo-config>`__ by specifying one or more configuration
+files on the command line. Each file is located and processed in order. When
+conflicting directives are provided in different files, the value in the last
+file processed takes precedence. List-valued directives (such as
+``exclude-path-patterns``) that are present in multiple files are concatenated.
+
+.. _configuration-discovery:
+
+Configuration Discovery
+-----------------------
+
+``tartufo`` uses two additional heuristics to locate specified configuration files.
+First, the current working directory is used as a base for relative filenames.
+Second, if the specified file does not exist in the specified directory, ``tartufo``
+will search upwards in the directory hierarchy looking for a file with the same
+name.
+
+Additionally, ``tartufo`` will look for a configuration file in the scan target
+repository or folder. It looks first for ``tartufo.toml``, and if that does not
+exist, then ``pyproject.toml``. The latter is searched for as a matter of
+convenience for Python projects, such as ``tartufo`` itself. This file must
+exist in the repository root directory (or target folder). The discovered
+configuration file will be processed after configuration files specified on the
+command line, unless it was also specified on the command line -- in which case,
+it will not be read a second time.
+
+Therefore, while it normally a target's configuration file will override settings
+specified using ``--config`` on the command line, this behavior can be overridden
+by specifying the target's configuration explicitly first, and then the desired
+"master" configuration second.
+
+Consider:
+
+.. code-block:: shell
+
+   tartufo --config myconfig.toml scan-local-repo .
+
+``tartufo`` will look for ``myconfig.toml`` in the current directory, and then
+in the parent directory if it does not exist in ``.``, etc., and then look for
+either ``tartufo.toml`` or (if not found) ``pyproject.toml`` in the current
+directory (only) because that is the target of the scan. Directives in, say,
+``tartufo.toml`` would supersede settings in ``myconfig.toml``.
+
+For purposes of this scan, empty files are considered not to exist.
+
+However:
+
+.. code-block:: shell
+
+   tartufo --config tartufo.toml --config myconfig.toml scan-local-repo .
+
+will cause ``tartufo`` to read ``tartufo.toml`` in the current directory
+(assuming it exists, and in parent directories otherwise), and then ``myconfig.toml``
```

Comment:
I believe this is still correct (although we plan a PR before the next release that will unilaterally remove the traversal logic, which we believe is effectively unused and won't be missed).

If you pass a relative path via `--config`, that is sent to `config.read_pyproject_toml()`, which sends it (unmodified) to `config.load_config_from_path()` (with `config_path` set to `.` and  `traverse` implicitly `True`).

If it can't find a file, `load_config_from_path()` recursively calls itself, except this time `config_path` is set to the parent of the original path (i.e. `../`) and repeats this until you reach the root directory, at which point `config_path.parent` matches `config_path` and the loop terminates.

Reply:
Okay, thanks. Yeah it sound like it will become immaterial soon, but just for completion: my understanding of the current/old behavior is: If you pass a relative path via `--config` , the click library transforms it immediately into an absolute path. The rest appears correct, except it is effectively unable to traverse due to that transformation that has already been done.